pkg/controller/nodelifecycle/node_lifecycle_controller.go

```diff
@@ -1206,6 +1273,89 @@ func (nc *Controller) handleDisruption(zoneToNodeConditions map[string][]*v1.Nod
 	}
 }
 
+func (nc *Controller) podUpdated(oldPod, newPod *v1.Pod) {
+	if newPod == nil {
+		return
+	}
+	if len(newPod.Spec.NodeName) != 0 && (oldPod == nil || newPod.Spec.NodeName != oldPod.Spec.NodeName) {
+		podItem := podUpdateItem{newPod.Namespace, newPod.Name}
+		nc.podUpdateQueue.Add(podItem)
+	}
+}
+
+func (nc *Controller) doPodProcessingWorker() {
+	for {
+		obj, shutdown := nc.podUpdateQueue.Get()
+		// "podUpdateQueue" will be shutdown when "stopCh" closed;
+		// we do not need to re-check "stopCh" again.
+		if shutdown {
+			return
+		}
+
+		podItem := obj.(podUpdateItem)
+		nc.processPod(podItem)
+	}
+}
+
+// processPod is processing events of assigning pods to nodes. In particular:
+// 1. for NodeReady=true node, taint eviction for this pod will be cancelled
+// 2. for NodeReady=false or unknown node, taint eviction of pod will happen and pod will be marked as not ready
+// 3. if node doesn't exist in cache, it will be skipped and handled later by doEvictionPass
+func (nc *Controller) processPod(podItem podUpdateItem) {
+	defer nc.podUpdateQueue.Done(podItem)
+	pod, err := nc.podLister.Pods(podItem.namespace).Get(podItem.name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// If the pod was deleted, there is no need to requeue.
+			return
+		}
+		klog.Warningf("Failed to read pod %v/%v: %v.", podItem.namespace, podItem.name, err)
+		nc.podUpdateQueue.AddRateLimited(podItem)
+		return
+	}
+
+	nodeName := pod.Spec.NodeName
+
+	nodeHealth := nc.nodeHealthMap.getDeepCopy(nodeName)
+	if nodeHealth == nil {
+		// Node data is not gathered yet or node has beed removed in the meantime.
+		// Pod will be handled by doEvictionPass method.
+		return
+	}
+
+	node, err := nc.nodeLister.Get(nodeName)
+	if err != nil {
+		klog.Warningf("Failed to read node %v: %v.", nodeName, err)
+		nc.podUpdateQueue.AddRateLimited(podItem)
+		return
+	}
+
+	_, currentReadyCondition := nodeutil.GetNodeCondition(nodeHealth.status, v1.NodeReady)
+	if currentReadyCondition == nil {
+		// Lack of NodeReady condition may only happen after node addition (or if it will be maliciously deleted).
+		// In both cases, the pod will be handled correctly (evicted if needed) during processing
+		// of the next node update event.
+		return
+	}
+
+	pods := []*v1.Pod{pod}
+	// In taint-based eviction mode, only node updates are processed by NodeLifecycleController.
+	// Pods are processed by TaintManager.
+	if !nc.useTaintBasedEvictions {
+		if err := nc.processNoTaintBaseEviction(node, currentReadyCondition, nc.nodeMonitorGracePeriod, pods); err != nil {
+			klog.Warningf("Unable to process pod %+v eviction from node %v: %v.", podItem, nodeName, err)
+			nc.podUpdateQueue.AddRateLimited(podItem)
```

Comment:
return ?

Reply:
It's additional pod status update call vs marking pod not ready in next pass.
I assume that it is very likely that if pod eviction failed than marking pods as not ready will fail as well... `return` added.